### ISSUE TITLE
Question: Low performance with Native Federation

### ISSUE THREAD

**Issue by ips219** [angular-architects/module-federation-plugin#559]
After moving from webpack to native federation I've noticed that surprisingly I've worst perfomance in local dev server with vite and esbuild than previously with webpack.

I am making some tests and trigering the rebuild with an small change takes a long time to complete.

`
 Changes detected. Rebuilding...
{
  "added": [],
  "modified": [.......],
  "removed": []

No output file changes.

Application bundle generation complete. [1.258 seconds]

 INFO  Building federation artefacts
 INFO  Done!

`

The bundle generation **takes just 1.2  seconds**, but the "Building federation artefacts" **takes almost 35 seconds** each time I make any change.

Considering my previous branch wich is currently working with Angular17 and webpack module federation it takes 2.4 seconds
![image](https://github.com/angular-architects/module-federation-plugin/assets/3368591/72e70d33-bee9-4a3f-9cf2-cf1d681659ba)

I guess I've some configuration wrong, but I've double checked the migration notes and the example and my angular.json is almost the same.
https://github.com/angular-architects/module-federation-plugin/blob/main/libs/native-federation/migrate-appbuilder.md
https://github.com/manfredsteyer/module-federation-plugin-example/blob/nf-standalone-starter/angular.json

Did anybody faced something similar?

Best regards.

**Comment by ips219** [comment]
Adding some additional details, after digging in the builder...

Checking the method which is taking so long, build-for-federation:buildForFederation() it seems that most of the time is taken in the bundle-exposed-and-mappings:bundleExposedAndMappings() and bundle-shared: bundleShared().

Both of them internally call same method build-utils:bundle() which the one that takes 35/40 seconds to execute both calls.

**Comment by manfredsteyer** [comment]
Thanks for this info.

Is it about ng serve or ng build?
Can you please share you package.json and your federation.config.js?

**Comment by ips219** [comment]
Sure...

Thanks for your support...

[info.zip](https://github.com/angular-architects/module-federation-plugin/files/15303772/info.zip)

I've seen in the esbuild documentation that antivirus may affect the bundle process... I am not sure if this can be the source of the problem. Just in in case I will test also disabling the AV.

**Comment by ips219** [comment]
After disabling the AV the performance improved a little bit but the bundle process still takes a long time to complete >20s (i added some logs).

`
Application bundle generation complete. [0.740 seconds]

 INFO    Calling bundleExposedAndMappings: Tue May 14 2024 10:14:35 GMT+0200
 INFO      Called bundleExposedAndMappings Tue May 14 2024 10:14:35 GMT+0200
 INFO      Calling bundle with 1 entrypoints Tue May 14 2024 10:14:35 GMT+0200
 INFO      Calling createBuildResultMap with 0 shared Tue May 14 2024 10:14:50 GMT+0200
 INFO      Adding 0 exposes Tue May 14 2024 10:14:50 GMT+0200
 INFO      Finished bundleExposedAndMappings Tue May 14 2024 10:14:50 GMT+0200
 INFO    Calling describeExposed: Tue May 14 2024 10:14:50 GMT+0200
 INFO    Calling bundleShared: Tue May 14 2024 10:14:50 GMT+0200
 INFO    Called bundleShared Tue May 14 2024 10:14:50 GMT+0200
 INFO    Calling bundle Tue May 14 2024 10:14:51 GMT+0200
 INFO    Bundle finished Tue May 14 2024 10:15:01 GMT+0200
 INFO    Files copied Tue May 14 2024 10:15:02 GMT+0200
 INFO    Calling describeSharedMappings: Tue May 14 2024 10:15:02 GMT+0200
 INFO   Calling writeFederationInfo: Tue May 14 2024 10:15:02 GMT+0200
 INFO    Calling writeImportMap: Tue May 14 2024 10:15:02 GMT+0200
 INFO  BuildFederation completed: Tue May 14 2024 10:15:02 GMT+0200
 INFO  Done!

`

**Comment by manfredsteyer** [comment]
Thanks. I will try it out with your dependencies.

Just to make sure we are talking about the same: Do we speak about the first ng serve/ ng build or subsequent ng serve/ ng build calls?

The first time they are called, it indeed takes a bit longer because all the deps need to be compiled for federation. The result of this is cached in node_modules/.cache. After that, it should really be fast as the deps are not compiled anymore.

**Comment by ips219** [comment]
I meant the subsequent calls...

I understand that the first ng serve takes longer as you said (cache generation...) but subsequent calls I expected to be lighweight.

In mfe + webpack once the dev server is running (ng serve) it takes up to 3s, now in mfe + native it is taking up to 35s to rebuild each developer change.

Best regards.

**Comment by dipu7388** [comment]
@manfredsteyer @jigfox   we have large repository which also  mono-repo library, using native federation build time for development/ production increased most time is consumed in **Building federation artefacts**.  its taking around 1-2 minutes.
![Uploading image.png…]()

**Comment by manfredsteyer** [comment]
Yes, that's possible, because building federation artifacts is an additional build step. 

But it should not take longer than the original Angular-build (37 sec in this case). Can you try to nuke your .angular folder and compare both times again?

**Comment by ladler1** [comment]
@ips219 could you solve the problem? I have a similar problem with long build times of federation artifacts

**Comment by livingplant** [comment]
We have the same issue. 
In a nx Monorepo we have a angular build Time for a Microfrontend of 1-2 Seconds with vite and the federation artifacts take 10-20 Seconds. It also uses way more Ressources. 
Has anyone solved this?

**Comment by rribeiro93k** [comment]
I have the same problem here, federation build time is too slow compared to angular build process. @manfredsteyer about this issue, we have some fix to solve? or still in analysis?

**Comment by livingplant** [comment]
Could this PR solve the Problem? https://github.com/angular-architects/module-federation-plugin/pull/588

**Comment by rribeiro93k** [comment]
@manfredsteyer @rainerhahnekamp any news or suggestion?

**Comment by ips219** [comment]
Just for the records... I migrated from NG17 to NG19 and I made a new attempt migrating from  webpack to native-generation... 

But... same results as one year ago. It takes 30/40 seconds **in the bundle generation**... In my case I cannot consider have different set of scripts... the application will be very dififcult to maintain because of the different federated modules... I have not a mono-repo design and it will be a pain.

So for the moment I've to keep webpack (I hope it will not become deprecated in a near future).

**Comment by SUSHEELsikarwar** [comment]
Facing Same issue.

**Comment by dr460nf1r3** [comment]
> Yes, that's possible, because building federation artifacts is an additional build step.
> 
> But it should not take longer than the original Angular-build (37 sec in this case). Can you try to nuke your .angular folder and compare both times again?

It definitely *does* take longer, *much* longer than the original Angular build. 
This is not a question but a very valid issue. At my work place, we've been checking this issue for news since ages.

**Comment by methode-user-1** [comment]
@manfredsteyer any news on this?

**Comment by methode-user-1** [comment]
@manfredsteyer any news on this?

**Comment by methode-user-1** [comment]
@manfredsteyer  Could we at least get a reply? I believe this is a serious bug, and it’s affecting many users.

**Comment by thanh-nguyen-codeleap** [comment]
@manfredsteyer Hi, is there any updates yet? This is a nightmare to develop anything meaningful with this library

**Comment by methode-user-1** [comment]
@manfredsteyer any news on this?

**Comment by rribeiro93k** [comment]
@manfredsteyer @rainerhahnekamp any suggestion on this?

**Comment by Aukevanoost** [comment]
Hi everyone, I'll have to read up and dig into this as well!  I wouldn't mind though. But this sounds like a complex issue to tackle so it would take some time since I'm relatively new.

It would help if anyone had a small reproducible repository where the main issue is clearly visible. It doesn't need to be big but that would help me isolate the issue. 

Cheers :)

**Comment by rainerhahnekamp** [comment]
@Aukevanoost, I guess it just the build speed of native federation. Nothing specific but more of a general issue.

**Comment by Aukevanoost** [comment]
Actually I've parallelized the single bundling tasks recently in #955 (it's in 20.1.7). But this isn't the main issue since most dependencies are handled in one go in the "bundleShared" step. 

Perhaps we can improve that further indeed!

**Comment by Aukevanoost** [comment]
Hi @fabianvieri. Do you mind setting verbose to true and sharing the logs? Then we can see what's taking so long.
